_posts/2018-10-05-a-macroscopic-take-on-metaprogramming.md

```diff
@@ -0,0 +1,293 @@
+---
+layout: post
+title:  "A macroscopic take on metaprogramming"
+date:   2018-10-05 17:00:00 +0100
+categories: elixir ruby javascript
+author: Natacha Segala
+---
+If you are as keen on DIY as I am and equally as helpless, the first time you're presented with some sort of power tool you may think 'This is scary, how could I possibly use this?'.
+
+In some respects, metaprogramming can feel just like this.
+
+A quick look at [wikipedia's definition](https://en.wikipedia.org/wiki/Metaprogramming){:target= "_blank"} gives us:
+"Metaprogramming is a programming technique in which computer programs have the ability to treat programs as their data."
+
+Another common definition is "writing code that writes code".
+
+One of the main unknowns when coding is user data.
+Everyday code does not necessarily solve unique problems.
+As developers, we use well defined conventions and set of functions we are familiar with to process this data and return whatever is defined in the product specifications.
+
+There should be room for abstraction and generalization and yet, the consensus is to try and avoid metaprogramming as much as possible.
+
+The prefix "meta" is self-referential which tends to indicate linguistically that metaprogramming encompasses all of programming including "normal" programming.
+We could think of "normal" programming as our ability to solve unique problems.
+All of the other kinds of general problems could – in theory – be handled by metaprogramming.
+
+Getting to know the reflective methods and tools every language we use offers is a way of refining one's craft.
+It also offers a glimpse into and a better understanding of the internals of the language.
+
+## Ruby
+
+When working with Ruby, chances are you will come across the DRY (don't repeat yourself) paradigm.
+
+The idea of avoiding duplication seems like a sound one in a more general context too.
+It will make any refactoring (or change) easier.
+
+Not a cure-all but definitely a tool to keep handy in one's dev belt.
+
+### Example
+
+We have a test suite responsible for checking the index page of our website includes the content we expect:
+
+```
+class IndexTest < TestCase
+  test_homepage_has_about_content do
+    visit '/'
+    assert page.has_content?(about)
+  end
+
+  test_homepage_has_news_content do
+    visit '/'
+    assert page.has_content?(news)
+  end
+
+  ...
+end
+```
+
+A quick online search or pairing could lead to reducing the code to:
+
+```
+class IndexTest < TestCase
+  HOMEPAGE_CONTENT = ['About', 'Meetups', 'Tweets', 'News'].freeze
+
+  HOMEPAGE_CONTENT.each do |content|
+    define_method("test_homepage_has_#{content}_content") do
+      visit '/'
+      assert page.has_content?(content)
+    end
+  end
+end
+```
+
+What does define_method do exactly and where does it come from?
```

Comment:
wrap ``define_method`` in backticks for syntax.

Reply:
Or even with a styled link [`define_method`](https://ruby-doc.org/core-2.2.0/Module.html#method-i-define_method)